src/v/serde/serde.h

```diff
@@ -166,6 +171,20 @@ inline constexpr auto const are_bytes_and_string_different = !(
 template<typename T>
 void write(iobuf&, T);
 
+template<class R, class P>
+bool check_duration_overflow(const std::chrono::duration<R, P>& duration) {
+    // Here we check the duration overflow by casting it into a wider type
+    // (double).
+    using wide_nano_s
+      = std::chrono::duration<double, std::chrono::nanoseconds::period>;
```

Comment:
I'm not sure what this means. IIUC it's casting into fractional seconds?

Reply:
This is for checking overflows in the duration (see the comment in L366). Since we are serializing as nanoseconds, the max duration (in nanoseconds) we support is nanoseconds::max(). Now std::chrono is weird in this regard, since nanoseconds and milliseconds are both represented by int64_t, nanoseconds::max() == milliseconds::max() == int64_max. Now for example one tries to cast milliseconds::max() into a nanosecond duration, that overflows silently.  We are avoiding that here and throwing if thats the case.
The way we detect the overflow is by casting it into a type that has more bits than  int64_t (which is double) and see if its > nano::max().